---
title: Zpracování chyb vstupně-výstupní operace v rozhraní .NET
ms.date: 08/27/2018
ms.technology: dotnet-standard
dev_langs:
- csharp
- vb
helpviewer_keywords:
- I/O, exception handling
- I/O, errors
author: rpetrusha
ms.author: ronpet
ms.workload:
- dotnet
- dotnetcore
ms.openlocfilehash: d2ff4e69596e721f485d107317f261231615c5a6
ms.sourcegitcommit: 9b552addadfb57fab0b9e7852ed4f1f1b8a42f8e
ms.translationtype: MT
ms.contentlocale: cs-CZ
ms.lasthandoff: 04/23/2019
ms.locfileid: "61752344"
---
# <a name="handling-io-errors-in-net"></a><span data-ttu-id="f89c9-102">Zpracování chyb vstupně-výstupní operace v rozhraní .NET</span><span class="sxs-lookup"><span data-stu-id="f89c9-102">Handling I/O errors in .NET</span></span>

<span data-ttu-id="f89c9-103">Kromě výjimky, které mohou být vyvolány v žádném volání metody (například <xref:System.OutOfMemoryException> systému je třeba zdůraznit, nebo s <xref:System.NullReferenceException> z důvodu chyby programátor), .NET metody systému souborů může vyvolat následující výjimky:</span><span class="sxs-lookup"><span data-stu-id="f89c9-103">In addition to the exceptions that can be thrown in any method call (such as an <xref:System.OutOfMemoryException> when a system is stressed or an <xref:System.NullReferenceException> due to programmer error), .NET file system methods can throw the following exceptions:</span></span>

- <span data-ttu-id="f89c9-104"><xref:System.IO.IOException?displayProperty=nameWithType>, základní třída všech <xref:System.IO> typy výjimek.</span><span class="sxs-lookup"><span data-stu-id="f89c9-104"><xref:System.IO.IOException?displayProperty=nameWithType>, the base class of all <xref:System.IO> exception types.</span></span> <span data-ttu-id="f89c9-105">Je vyvolána výjimka, chyby, jehož návratové kódy z operačního systému není mapují přímo na jiný typ výjimky.</span><span class="sxs-lookup"><span data-stu-id="f89c9-105">It is thrown for errors whose return codes from the operating system don't directly map to any other exception type.</span></span>
- <span data-ttu-id="f89c9-106"><xref:System.IO.FileNotFoundException?displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-106"><xref:System.IO.FileNotFoundException?displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-107"><xref:System.IO.DirectoryNotFoundException?displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-107"><xref:System.IO.DirectoryNotFoundException?displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-108"><xref:System.IO.DriveNotFoundException??displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-108"><xref:System.IO.DriveNotFoundException??displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-109"><xref:System.IO.PathTooLongException?displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-109"><xref:System.IO.PathTooLongException?displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-110"><xref:System.OperationCanceledException?displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-110"><xref:System.OperationCanceledException?displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-111"><xref:System.UnauthorizedAccessException?displayProperty=nameWithType>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-111"><xref:System.UnauthorizedAccessException?displayProperty=nameWithType>.</span></span>
- <span data-ttu-id="f89c9-112"><xref:System.ArgumentException?displayProperty=nameWithType>, která je vyvolána neplatná cesta znaků v rozhraní .NET Framework a .NET Core 2.0 a předchozí verze.</span><span class="sxs-lookup"><span data-stu-id="f89c9-112"><xref:System.ArgumentException?displayProperty=nameWithType>, which is thrown for invalid path characters on .NET Framework and on .NET Core 2.0 and previous versions.</span></span>
- <span data-ttu-id="f89c9-113"><xref:System.NotSupportedException?displayProperty=nameWithType>, která je vyvolána neplatná použití dvojteček v rozhraní .NET Framework.</span><span class="sxs-lookup"><span data-stu-id="f89c9-113"><xref:System.NotSupportedException?displayProperty=nameWithType>, which is thrown for invalid colons in .NET Framework.</span></span>
- <span data-ttu-id="f89c9-114"><xref:System.Security.SecurityException?displayProperty=nameWithType>, která je vyvolána pro aplikace běžící v omezené vztahu důvěryhodnosti, které nemají potřebná oprávnění na rozhraní .NET Framework pouze.</span><span class="sxs-lookup"><span data-stu-id="f89c9-114"><xref:System.Security.SecurityException?displayProperty=nameWithType>, which is thrown for applications running in limited trust that lack the necessary permissions on .NET Framework only.</span></span> <span data-ttu-id="f89c9-115">(Úplný vztah důvěryhodnosti je výchozí hodnota v rozhraní .NET Framework.)</span><span class="sxs-lookup"><span data-stu-id="f89c9-115">(Full trust is the default on .NET Framework.)</span></span>

## <a name="mapping-error-codes-to-exceptions"></a><span data-ttu-id="f89c9-116">Mapování kódů chyb a výjimek</span><span class="sxs-lookup"><span data-stu-id="f89c9-116">Mapping error codes to exceptions</span></span>

<span data-ttu-id="f89c9-117">Protože systém souborů je prostředek operačního systému, vstupně-výstupní metody v .NET Core a .NET Framework obalují volání základního operačního systému.</span><span class="sxs-lookup"><span data-stu-id="f89c9-117">Because the file system is an operating system resource, I/O methods in both .NET Core and .NET Framework wrap calls to the underlying operating system.</span></span> <span data-ttu-id="f89c9-118">Když dojde k chybě vstupně-výstupní operace v kódu prováděném operačním systémem, operační systém vrátí informace o chybě metodě .NET vstupně-výstupních operací.</span><span class="sxs-lookup"><span data-stu-id="f89c9-118">When an I/O error occurs in code executed by the operating system, the operating system returns error information to the .NET I/O method.</span></span> <span data-ttu-id="f89c9-119">Metoda pak přeloží informace o chybě, obvykle v podobě chybový kód, na typ .NET výjimky.</span><span class="sxs-lookup"><span data-stu-id="f89c9-119">The method then translates the error information, typically in the form of an error code, into a .NET exception type.</span></span> <span data-ttu-id="f89c9-120">Ve většině případů to dělá tak, že přímo překládá na jeho odpovídající typ výjimky; kód chyby: neprovede žádné speciální mapování chybu na základě kontextu volání metody.</span><span class="sxs-lookup"><span data-stu-id="f89c9-120">In most cases, it does this by directly translating the error code into its corresponding exception type; it does not perform any special mapping of the error based on the context of the method call.</span></span>

<span data-ttu-id="f89c9-121">Například v operačním systému Windows, volání metody, která vrací kód chyby `ERROR_FILE_NOT_FOUND` (nebo 0x02) mapuje na <xref:System.IO.FileNotFoundException>a kód chyby `ERROR_PATH_NOT_FOUND` (nebo 0x03) se mapuje <xref:System.IO.DirectoryNotFoundException>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-121">For example, on the Windows operating system, a method call that returns an error code of `ERROR_FILE_NOT_FOUND` (or 0x02) maps to a <xref:System.IO.FileNotFoundException>, and an error code of `ERROR_PATH_NOT_FOUND` (or 0x03) maps to a <xref:System.IO.DirectoryNotFoundException>.</span></span>

<span data-ttu-id="f89c9-122">Přesné podmínky, za kterých operační systém vrátí určité chybové kódy je však často nedokumentované nebo špatně dokument.</span><span class="sxs-lookup"><span data-stu-id="f89c9-122">However, the precise conditions under which the operating system returns particular error codes is often undocumented or poorly documented.</span></span> <span data-ttu-id="f89c9-123">V důsledku toho může dojít k neočekávaným výjimkám.</span><span class="sxs-lookup"><span data-stu-id="f89c9-123">As a result, unexpected exceptions can occur.</span></span> <span data-ttu-id="f89c9-124">Například vzhledem k tomu, že pracujete s adresáři, nikoli soubor, které by uživatel očekával poskytující neplatná cesta k adresáři na <xref:System.IO.DirectoryInfo.%23ctor%2A?displayProperty=nameWithType> vyvolá konstruktor <xref:System.IO.DirectoryNotFoundException>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-124">For example, because you are working with a directory rather than a file, you would expect that providing an invalid directory path to the <xref:System.IO.DirectoryInfo.%23ctor%2A?displayProperty=nameWithType> constructor throws a <xref:System.IO.DirectoryNotFoundException>.</span></span> <span data-ttu-id="f89c9-125">Ale může také vyvolat <xref:System.IO.FileNotFoundException>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-125">However, it may also throw a <xref:System.IO.FileNotFoundException>.</span></span>

## <a name="exception-handling-in-io-operations"></a><span data-ttu-id="f89c9-126">Zpracování výjimek v vstupně-výstupních operací</span><span class="sxs-lookup"><span data-stu-id="f89c9-126">Exception handling in I/O operations</span></span>

<span data-ttu-id="f89c9-127">Z důvodu této závislosti na operačním systému podmínky výjimek identické (například adresář nebyl nalezen chyba v našem příkladu) může vést k metodu vstupně-výstupních operací vyvolání některý celou třídu výjimky vstupně-výstupních operací.</span><span class="sxs-lookup"><span data-stu-id="f89c9-127">Because of this reliance on the operating system, identical exception conditions (such as the directory not found error in our example) can result in an I/O method throwing any one of the entire class of I/O exceptions.</span></span> <span data-ttu-id="f89c9-128">To znamená, že při volání rozhraní API pro vstupně-výstupních operací, váš kód by měla být připravena zpracovat většinu nebo všechny tyto výjimky, jak je znázorněno v následující tabulce:</span><span class="sxs-lookup"><span data-stu-id="f89c9-128">This means that, when calling I/O APIs, your code should be prepared to handle most or all of these exceptions, as shown in the following table:</span></span>

| <span data-ttu-id="f89c9-129">Typ výjimky</span><span class="sxs-lookup"><span data-stu-id="f89c9-129">Exception type</span></span> | <span data-ttu-id="f89c9-130">.NET Core</span><span class="sxs-lookup"><span data-stu-id="f89c9-130">.NET Core</span></span> | <span data-ttu-id="f89c9-131">.NET Framework</span><span class="sxs-lookup"><span data-stu-id="f89c9-131">.NET Framework</span></span> |
|---|---|---|
| <xref:System.IO.IOException> | <span data-ttu-id="f89c9-132">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-132">Yes</span></span> | <span data-ttu-id="f89c9-133">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-133">Yes</span></span> |
| <xref:System.IO.FileNotFoundException> | <span data-ttu-id="f89c9-134">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-134">Yes</span></span> | <span data-ttu-id="f89c9-135">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-135">Yes</span></span> |
| <xref:System.IO.DirectoryNotFoundException> | <span data-ttu-id="f89c9-136">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-136">Yes</span></span> | <span data-ttu-id="f89c9-137">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-137">Yes</span></span> |
| <xref:System.IO.DriveNotFoundException?> | <span data-ttu-id="f89c9-138">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-138">Yes</span></span> | <span data-ttu-id="f89c9-139">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-139">Yes</span></span> |
| <xref:System.IO.PathTooLongException> | <span data-ttu-id="f89c9-140">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-140">Yes</span></span> | <span data-ttu-id="f89c9-141">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-141">Yes</span></span> |
| <xref:System.OperationCanceledException> | <span data-ttu-id="f89c9-142">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-142">Yes</span></span> | <span data-ttu-id="f89c9-143">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-143">Yes</span></span> |
| <xref:System.UnauthorizedAccessException> | <span data-ttu-id="f89c9-144">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-144">Yes</span></span> | <span data-ttu-id="f89c9-145">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-145">Yes</span></span> |
| <xref:System.ArgumentException> | <span data-ttu-id="f89c9-146">.NET core 2.0 a starší</span><span class="sxs-lookup"><span data-stu-id="f89c9-146">.NET Core 2.0 and earlier</span></span>| <span data-ttu-id="f89c9-147">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-147">Yes</span></span> |
| <xref:System.NotSupportedException> | <span data-ttu-id="f89c9-148">Ne</span><span class="sxs-lookup"><span data-stu-id="f89c9-148">No</span></span> | <span data-ttu-id="f89c9-149">Ano</span><span class="sxs-lookup"><span data-stu-id="f89c9-149">Yes</span></span> |
| <xref:System.Security.SecurityException> | <span data-ttu-id="f89c9-150">Ne</span><span class="sxs-lookup"><span data-stu-id="f89c9-150">No</span></span> | <span data-ttu-id="f89c9-151">Pouze omezený vztah důvěryhodnosti</span><span class="sxs-lookup"><span data-stu-id="f89c9-151">Limited trust only</span></span> |

## <a name="handling-ioexception"></a><span data-ttu-id="f89c9-152">IOException – zpracování</span><span class="sxs-lookup"><span data-stu-id="f89c9-152">Handling IOException</span></span>

<span data-ttu-id="f89c9-153">Jako základní třída pro výjimky ve <xref:System.IO> obor názvů, <xref:System.IO.IOException> je také pro libovolný kód chyby, které nejsou namapované na typ předdefinovaná výjimka vyvolána.</span><span class="sxs-lookup"><span data-stu-id="f89c9-153">As the base class for exceptions in the <xref:System.IO> namespace, <xref:System.IO.IOException> is also thrown for any error code that does not map to a predefined exception type.</span></span> <span data-ttu-id="f89c9-154">To znamená, že mohou být vyvolány ve všech vstupně-výstupní operace.</span><span class="sxs-lookup"><span data-stu-id="f89c9-154">This means that it can be thrown by any I/O operation.</span></span>

> [!IMPORTANT]
> <span data-ttu-id="f89c9-155">Protože <xref:System.IO.IOException> je základní třídou jiné typy výjimek v <xref:System.IO> obor názvů, můžete pracovat v `catch` blokovat po jsme zpracování I O souvisejícím s/výjimky.</span><span class="sxs-lookup"><span data-stu-id="f89c9-155">Because <xref:System.IO.IOException> is the base class of the other exception types in the <xref:System.IO> namespace, you should handle in a `catch` block after you've handled the other I/O-related exceptions.</span></span>

<span data-ttu-id="f89c9-156">Kromě toho počínaje .NET Core 2.1, ověřovacích kontrol správnosti cestu (například k zajištění, že neplatné znaky nejsou k dispozici v cestě) byly odebrány, a modul runtime vyvolá výjimku, mapovaná z kód chyby operačního systému místo z vlastního ověřovacího kódu.</span><span class="sxs-lookup"><span data-stu-id="f89c9-156">In addition, starting with .NET Core 2.1, validation checks for path correctness (for example, to ensure that invalid characters are not present in a path) have been removed, and the runtime throws an exception mapped from an operating system error code rather than from its own validation code.</span></span> <span data-ttu-id="f89c9-157">V tomto případě je pravděpodobně výjimka, která má být vyvolána <xref:System.IO.IOException>, i když může být vyvoláno také jakýkoli jiný typ výjimky.</span><span class="sxs-lookup"><span data-stu-id="f89c9-157">The most likely exception to be thrown in this case is an <xref:System.IO.IOException>, although any other exception type could also be thrown.</span></span>

<span data-ttu-id="f89c9-158">Všimněte si, že v váš kód zpracování výjimek, vždy zpracovávejte <xref:System.IO.IOException> poslední.</span><span class="sxs-lookup"><span data-stu-id="f89c9-158">Note that, in your exception handling code, you should always handle the <xref:System.IO.IOException> last.</span></span> <span data-ttu-id="f89c9-159">Jinak protože je základní třídy pro všechny ostatní výjimky vstupně-výstupních operací, nebude vyhodnocen bloky catch odvozených tříd.</span><span class="sxs-lookup"><span data-stu-id="f89c9-159">Otherwise, because it is the base class of all other IO exceptions, the catch blocks of derived classes will not be evaluated.</span></span>

<span data-ttu-id="f89c9-160">V případě třídy <xref:System.IO.IOException>, získáte další informace o chybě z [IOException.HResult](xref:System.Exception.HResult) vlastnost.</span><span class="sxs-lookup"><span data-stu-id="f89c9-160">In the case of an <xref:System.IO.IOException>, you can get additional error information from the [IOException.HResult](xref:System.Exception.HResult) property.</span></span> <span data-ttu-id="f89c9-161">Převede hodnotu HResult kódu chyby Win32, můžete odstranit horní 16 bitů 32bitová hodnota.</span><span class="sxs-lookup"><span data-stu-id="f89c9-161">To convert the HResult value to a Win32 error code, you strip out the upper 16 bits of the 32-bit value.</span></span> <span data-ttu-id="f89c9-162">V následující tabulce jsou uvedeny kódy chyb, které mohou být uzavřen do <xref:System.IO.IOException>.</span><span class="sxs-lookup"><span data-stu-id="f89c9-162">The following table lists error codes that may be wrapped in an <xref:System.IO.IOException>.</span></span>

| <span data-ttu-id="f89c9-163">Hodnota HResult</span><span class="sxs-lookup"><span data-stu-id="f89c9-163">HResult</span></span> | <span data-ttu-id="f89c9-164">Konstanta</span><span class="sxs-lookup"><span data-stu-id="f89c9-164">Constant</span></span> | <span data-ttu-id="f89c9-165">Popis</span><span class="sxs-lookup"><span data-stu-id="f89c9-165">Description</span></span> |
| --- | --- | --- |
| <span data-ttu-id="f89c9-166">ERROR_SHARING_VIOLATION</span><span class="sxs-lookup"><span data-stu-id="f89c9-166">ERROR_SHARING_VIOLATION</span></span> | <span data-ttu-id="f89c9-167">32</span><span class="sxs-lookup"><span data-stu-id="f89c9-167">32</span></span> | <span data-ttu-id="f89c9-168">Chybí název souboru nebo souboru nebo adresáře se používá.</span><span class="sxs-lookup"><span data-stu-id="f89c9-168">The file name is missing, or the file or directory is in use.</span></span> |
| <span data-ttu-id="f89c9-169">ERROR_FILE_EXISTS</span><span class="sxs-lookup"><span data-stu-id="f89c9-169">ERROR_FILE_EXISTS</span></span> | <span data-ttu-id="f89c9-170">80</span><span class="sxs-lookup"><span data-stu-id="f89c9-170">80</span></span> | <span data-ttu-id="f89c9-171">Soubor již existuje.</span><span class="sxs-lookup"><span data-stu-id="f89c9-171">The file already exists.</span></span> |
| <span data-ttu-id="f89c9-172">ERROR_INVALID_PARAMETER</span><span class="sxs-lookup"><span data-stu-id="f89c9-172">ERROR_INVALID_PARAMETER</span></span> | <span data-ttu-id="f89c9-173">87</span><span class="sxs-lookup"><span data-stu-id="f89c9-173">87</span></span> | <span data-ttu-id="f89c9-174">Argument zadaný pro metodu je neplatný.</span><span class="sxs-lookup"><span data-stu-id="f89c9-174">An argument supplied to the method is invalid.</span></span> |
| <span data-ttu-id="f89c9-175">ERROR_ALREADY_EXISTS</span><span class="sxs-lookup"><span data-stu-id="f89c9-175">ERROR_ALREADY_EXISTS</span></span> | <span data-ttu-id="f89c9-176">183</span><span class="sxs-lookup"><span data-stu-id="f89c9-176">183</span></span> | <span data-ttu-id="f89c9-177">Soubor nebo adresář již existuje.</span><span class="sxs-lookup"><span data-stu-id="f89c9-177">The file or directory already exists.</span></span> |

<span data-ttu-id="f89c9-178">Můžete zpracovat pomocí `When` klauzulí catch příkazu, jak ukazuje následující příklad.</span><span class="sxs-lookup"><span data-stu-id="f89c9-178">You can handle these using a `When` clause in a catch statement, as the following example shows.</span></span>

[!code-csharp[io-exception-handling](~/samples/snippets/standard/io/io-exceptions/cs/io-exceptions.cs)]
[!code-vb[io-exception-handling](~/samples/snippets/standard/io/io-exceptions/vb/io-exceptions.vb)]

## <a name="see-also"></a><span data-ttu-id="f89c9-179">Viz také:</span><span class="sxs-lookup"><span data-stu-id="f89c9-179">See also</span></span>

- [<span data-ttu-id="f89c9-180">Zpracování a vyvolání výjimek v rozhraní .NET</span><span class="sxs-lookup"><span data-stu-id="f89c9-180">Handling and throwing exceptions in .NET</span></span>](../exceptions/index.md)
- [<span data-ttu-id="f89c9-181">Zpracování výjimek (Task Parallel Library)</span><span class="sxs-lookup"><span data-stu-id="f89c9-181">Exception handling (Task Parallel Library)</span></span>](../parallel-programming/exception-handling-task-parallel-library.md)
- [<span data-ttu-id="f89c9-182">Doporučené postupy pro výjimky</span><span class="sxs-lookup"><span data-stu-id="f89c9-182">Best practices for exceptions</span></span>](../exceptions/best-practices-for-exceptions.md)
- [<span data-ttu-id="f89c9-183">Postup používání specifických výjimek v bloku catch</span><span class="sxs-lookup"><span data-stu-id="f89c9-183">How to use specific exceptions in a catch block</span></span>](../exceptions/how-to-use-specific-exceptions-in-a-catch-block.md)
